### PR TITLE
Invoke postinstallation hooks

### LINKED ISSUE #82: Support for native modules/addons
It seems right now `ied` doesn't support node.js addons (native modules). When the user installs packages that have C/C++ native code, for example `libpq`, the installer just downloads the package and ignores the `bindings.gyp` file which specifies how the package should be compiled.

I assume supporting addons/native modules is is a planned feature, and if so I would like to contribute a PR. Can you point me in the right place to start?

### PR DESCRIPTION
**Pull request by just-boris**
closes #82
Also, this feature was asked in #43 as the necessary thing to install `electron-prebuilt`.

It is not ready to merge, yet, I'd like just share my results, and I want to add more tests before this will be merged.

And there is a thing with native packages. They require `node-gyp` to be installed. NPM carries bundled `node-gyp` command. To get this work with IED, we need to install it: `npm install -g node-gyp`. Maybe it should be mentioned in the docs.

### CODE REVIEW

**Comment by rauchg** [comment]
+1

**Comment by rauchg** [comment]
@just-boris Just tested this on a local project for which `ied` was failing due to the lack of `postinstall`, and it worked like a charm! Thank you

**Comment by alexanderGugel** [comment]
That's really great! Big fan of this!

File: lib/install_cmd.js [hunk 1]
```diff
@@ -3,9 +3,7 @@
 var path = require('path')
 var async = require('async')
 var assign = require('object-assign')
-var mkdirp = require('mkdirp')
 var expose = require('./expose')
-var config = require('./config')
```
Comment (just-boris) [inline lib/install_cmd.js]: doesn't used anymore, standard linter complains about that

**Comment by just-boris** [comment]
The tests are lying, it is still not finished

File: .travis.yml [hunk 2]
```diff
@@ -1,4 +1,13 @@
 language: node_js
+env:
+    - CXX=g++-4.8
+addons:
```
Comment (just-boris) [inline .travis.yml]: It is a necessary change, described in [Travis documentation](https://docs.travis-ci.com/user/languages/javascript-with-nodejs#Node.js-v4-(or-io.js-v3)-compiler-requirements)

**Comment by just-boris** [comment]
Fixed, the actual reason wasn't related to postinstall feature. I sent another pull request aimed to fix it: https://github.com/alexanderGugel/ied/pull/90

This PR is rebased and contains that fixes as well.

**Comment by just-boris** [comment]
@alexanderGugel it is ready to merge now :ship:

**Comment by alexanderGugel** [comment]
@just-boris Just merged the other PR. Will merge this one later today if possible. Thanks again!

File: .travis.yml (hunk 2 above)
Comment (alexanderGugel) [inline .travis.yml]: Gotcha. Looks good.

File: package.json [hunk 3]
```diff
@@ -33,7 +33,8 @@
     "sepia": "^2.0.1",
     "snazzy": "^2.0.1",
     "standard": "^5.3.1",
-    "tape": "^4.2.2"
+    "tape": "^4.2.2",
+    "tmp": "0.0.28"
```
Comment (alexanderGugel) [inline package.json]: Hmmm... wondering if we should also use this package for our actual application code... Looks really good to me (shouldn't be part of this PR though anyways).

File: lib/cache.js [hunk 4]
```diff
@@ -20,7 +20,9 @@ function read (uid) {
 
 exports.init = init
 function init (cb) {
-  mkdirp(path.join(config.cacheDir, '.tmp'), cb)
+  mkdirp(path.join(config.cacheDir, '.tmp'), function (err) {
```
Comment (alexanderGugel) [inline lib/cache.js]: This should already be on master now.

File: lib/install_cmd.js (hunk 1 above)
Comment (alexanderGugel) [inline lib/install_cmd.js]: Cool cool!

File: lib/run.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
+var path = require('path')
```
Comment (alexanderGugel) [inline lib/run.js]: Really minor thing: `run.js` should be added to `lib/index.js` to be exported.

File: package.json (hunk 3 above)
Comment (just-boris) [inline package.json]: FYI, that directory will be automatically deleted after process finish. Is it good behaviour for you?

Comment (alexanderGugel) [inline package.json]: We currently do that manually for downloading the package tarballs. Shouldn't be changed for now though. This is great.

**Comment by just-boris** [comment]
Well, got this, I am going to fix it on the next week, too many things to do at the current

**Comment by just-boris** [comment]
done

**Comment by just-boris** [comment]
@alexanderGugel ping

**Comment by STRML** [comment]
As `ied` is becoming a faster/higher security version of `npm`, might it make sense to offer a global flag to turn off `postinstall` hooks, or whitelist packages that are allowed to execute them?

**Comment by alexanderGugel** [comment]
@STRML Yes. postinstallation hooks (and lifecycle scripts in general) are in a sense a complete security fiasco IMO... I think they should actually be disabled by default....

I like this PR.

@just-boris Thanks for putting this together! Merged! Will be part of next release!

**Comment by vladwing** [comment]
@alexanderGugel if you disable life-cycle scripts, how can one install native modules? The person installing the npm module might not even know the module is half-installed.

**Comment by alexanderGugel** [comment]
@vladwing You would have to decide on a case by case basis whether or not you want to build those dependencies. It sucks, but the alternative is to just hope that none of those 300 maintainers that are behind the subdependencies or some package suddenly turn evil and decide to take over a thousand machines in a matter of hours. I'm surprised nobody has done that already.

**Comment by vladwing** [comment]
I totally agree it's wrong to allow code to run on your machine for no good reason. What I was suggesting is we need a way to alert that:

1. the module has a native module it needs built
2. the module has a native module it needs built and life-cycle scripts
3. the module has life-cycle scripts, but no native modules

In that case ied should not install that module unless the user specifically permits it by adding its hash(not the package name/version anything else) to a whitelist.

How does it sound?

**Comment by STRML** [comment]
While npm is still dominant, that strategy seems fine, but if ied ever
becomes popular, won't that just mean that a malware author has only a
slightly higher bar to entry?

The idea of expecting every dev to audit all hooks also seems untenable.
It's like we need a per-hash voting system.

**Comment by STRML** [comment]
OTOH, are there actually things that a postinstall hook can do that a
module can't do upon a require()?

**Comment by just-boris** [comment]
Let's move the discussion about it in the issue: https://github.com/alexanderGugel/ied/issues/99
